[ELY-405] LdapKeyStore

honza889 opened the pull request:
LDAP backed KeyStore.

Tested with *inetOrgPerson* attributes as defined in RFC 2798:
* userCertificate (PKCS#7) - certificate
* userSMIMECertificate (PKCS#7) - certificate chain
* userPKCS12 (PKCS#12) - keystore with key + chain, password protected

Custom builder is used, because standard KeyStore API doesnt allow to pass informations (like LDAP settings) into the KeyStoreSpi from outside.

darranl commented on src/main/java/org/wildfly/security/keystore/LdapKeyStoreSpi.java:
```diff
@@ -0,0 +1,443 @@
+
+import static org.wildfly.security._private.ElytronMessages.log;
+
+import org.apache.directory.api.util.GeneralizedTime;
```
Is there any way we can replace this so we don't depend on the Apache APIs within the main project?

darranl commented on pom.xml:
```diff
@@ -260,6 +260,12 @@
 
+        <dependency>
+            <groupId>org.apache.directory.api</groupId>
+            <artifactId>api-util</artifactId>
```
Have commented in the code but would be nice if we could avoid adding a dependency on the Apache APIs directly, especially if it is only for something small.

darranl commented on src/test/resources/ldap/elytron-keystore-tests.ldif:
```diff
@@ -0,0 +1,124 @@
+objectclass: organizationalUnit
+ou: keystore
+
+dn: cn=firefly,ou=keystore,dc=elytron,dc=wildfly,dc=org
```
Looking at the name I assume the content here is based on the Elytron certificate authority within the project, it may be nice if we could have a Readme near here to confirm this and some information on how to concert from the CA files to the LDIF so if things are re-generated it will not take much to update this as well.

darranl commented:
Couple of small comments but looks good - combined with the subsystem changes how would you feel about putting together a small blog post / article on this new feature?

honza889 commented:
* Rewritten parsing of LDAP timestamp, doesnt create dependency on Apache Directory API already
* Added bash script allowing generating of LDIF with cert+chain+key (ldap-keystore-gen.sh)

honza889 commented:
Hmm... ok, dont merge/review, I will have to replace DirContextFactory by Supplier<DirContext> jet.

honza889 commented:
Prepared for review now.
* DirContextFactory replaced by Supplier in LdapKeyStore too

darranl commented:
Looks good, thanks for the updates, added my +1.
